src/Data/Text.hs

```diff
@@ -1176,6 +1180,21 @@ minimum :: HasCallStack => Text -> Char
 minimum t = S.minimum (stream t)
 {-# INLINE minimum #-}
 
+-- TODO
+isAscii :: Text -> Bool
+isAscii (Text (A.ByteArray arr) off len) =
+    let arrPtr = Exts.Ptr (Exts.byteArrayContents# arr)
```

Comment:
This is unsafe/wrong. `Text`'s bytearray is not pinned, so it may move after you take a pointer. `ccall unsafe` is not enough.

You need to import `_hs_text_is_ascii` as taking `ByteArray#` argument (grep for `memchr`). Then there won't be a race condition.

Reply:
I thought that was wrong! Thanks very much for the help, I don't understand memory pinning & had hoped to get poked in the right place. So the FFI resolves `ByteArray#` and `Ptr Word8` to the same C type, but they do different things. Is this hidden in the compiler somewhere? *Edit: I failed to read a line in GHC.Exts about `UnliftedFFITypes`! There's even a table about usage safety! Awesome.*

If you don't mind, another thing I was couldn't figure out: what is the difference between an `IO a` and a "pure" `a` foreign import? (Either way it's the same C return type, right? just info for the compiler?)

Reply:
If the function is pure like `sin` then you don't need `IO`.

In

```haskell
foreign import ccall unsafe "_hs_text_is_ascii" c_is_ascii
    :: Ptr Word8 -> Ptr Word8 -> IO CSize
```

The C function reads memory, so that is not pure. (Another thread can write into the same memory range e.g.)

```haskell
foreign import ccall unsafe "_hs_text_is_ascii_2" c_is_ascii_2
    :: ByteArray# -> CSize -> IO CSize
```

is OTOH "pure" as `ByteArray#` is passed in "by value". That's why `IO` was dropped recently (See https://github.com/haskell/text/pull/482)

Reply:
I see! I've done your 2nd example without the `IO`, which calls the main function with pointers to the beginning and end of the data calculated (the same pattern is used in some `Data.Text.Array` foreign calls):

```haskell
isAscii (Text (A.ByteArray arr) (Exts.I# off#) (Exts.I# len#)) =
    cSizeToInt (c_is_ascii_2 arr off# len#) == (Exts.I# len#)
foreign import ccall unsafe "_hs_text_is_ascii_2" c_is_ascii_2
    :: ByteArray# -> Exts.Int# -> Exts.Int# -> CSize
```